packages/@tailwindcss-upgrade/src/codemods/template/signatures.ts

```diff
@@ -0,0 +1,404 @@
+import { substituteAtApply } from '../../../../tailwindcss/src/apply'
+import { atRule, styleRule, toCss, walk, type AstNode } from '../../../../tailwindcss/src/ast'
+import { printArbitraryValue } from '../../../../tailwindcss/src/candidate'
+import * as SelectorParser from '../../../../tailwindcss/src/compat/selector-parser'
+import type { DesignSystem } from '../../../../tailwindcss/src/design-system'
+import { ThemeOptions } from '../../../../tailwindcss/src/theme'
+import { DefaultMap } from '../../../../tailwindcss/src/utils/default-map'
+import * as ValueParser from '../../../../tailwindcss/src/value-parser'
+import { dimensions } from '../../utils/dimension'
+
+// Given a utility, compute a signature that represents the utility. The
+// signature will be a normalised form of the generated CSS for the utility, or
+// a unique symbol if the utility is not valid. The class in the selector will
+// be replaced with the `.x` selector.
+//
+// This function should only be passed the base utility so `flex`, `hover:flex`
+// and `focus:flex` will all use just `flex`. Variants are handled separately.
+//
+// E.g.:
+//
+// | UTILITY          | GENERATED SIGNATURE     |
+// | ---------------- | ----------------------- |
+// | `[display:flex]` | `.x { display: flex; }` |
+// | `flex`           | `.x { display: flex; }` |
+//
+// These produce the same signature, therefore they represent the same utility.
+export const computeUtilitySignature = new DefaultMap<
+  DesignSystem,
+  DefaultMap<string, string | Symbol>
+>((designSystem) => {
+  return new DefaultMap<string, string | Symbol>((utility) => {
+    try {
+      // Ensure the prefix is added to the utility if it is not already present.
+      utility =
+        designSystem.theme.prefix && !utility.startsWith(designSystem.theme.prefix)
+          ? `${designSystem.theme.prefix}:${utility}`
+          : utility
+
+      // Use `@apply` to normalize the selector to `.x`
+      let ast: AstNode[] = [styleRule('.x', [atRule('@apply', utility)])]
+
+      temporarilyDisableThemeInline(designSystem, () => substituteAtApply(ast, designSystem))
+
+      // We will be mutating the AST, so we need to clone it first to not affect
+      // the original AST
+      ast = structuredClone(ast)
+
+      // Optimize the AST. This is needed such that any internal intermediate
+      // nodes are gone. This will also cleanup declaration nodes with undefined
+      // values or `--tw-sort` declarations.
+      walk(ast, (node, { replaceWith }) => {
```

Comment:
Is there a way we can reuse the optimize method from core here? Feels weird to have to re-implement this and it'll likely mean we'll diverge eventually 🤔

Reply:
I originally did that, but it made things slower because it's creating a new AST and tracking CSS variables, removing it from the output, hoisting at-root, ... and all of that jazz. But you are right that this might result in diverged code. 🤔 

Maybe it's fine, but here are some numbers from projects I tested it on.

Running this on Tailwind Plus: (not a lot of arbitrary properties/values are used)
- With the current implementation: 48s
- With the `optimizeAst` step: 55s

Running this on all Tailwind UI Templates (ran the script in parallel for each template and this is the final result):
- With the current implementation: 1m21s
- With the `optimizeAst` step: 2m48s
 
Running this on a personal project:
- With the current implementation: 29s
- With the `optimizeAst` step: 1m8s

Reply:
Oof that's quite a difference. Yikes. Maybe `optimizeAst` needs to be refactored somehow? Idk, let's keep thinking about this, nothing you need to change immediately I feel like.

Reply:
Wow that's a massive difference. We should absolutely profile optimizeAst(…) then and see if we can do something to make it faster.

Reply:
It's the upgrade tool so we can get away with slower builds, but this was surprising to me as well haha